services/src/northbound/src/main/java/org/openkilda/northbound/messaging/kafka/KafkaMessagingChannel.java

```diff
@@ -0,0 +1,202 @@
+     */
+    @PostConstruct
+    public void setUp() {
+        pendingRequests = new PassiveExpiringMap<>(expiredTime, TimeUnit.MINUTES, new ConcurrentHashMap<>());
+        pendingChunkedRequests = new PassiveExpiringMap<>(expiredTime, TimeUnit.MINUTES, new ConcurrentHashMap<>());
+
+        unlinkedChunks = new PassiveExpiringMap<>(expiredTime, TimeUnit.MINUTES, new HashMap<>());
+        messagesChains = new PassiveExpiringMap<>(expiredTime, TimeUnit.MINUTES, new HashMap<>());
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public CompletableFuture<InfoData> sendAndGet(String topic, Message message) {
+        CompletableFuture<InfoData> future = new CompletableFuture<>();
+        pendingRequests.put(message.getCorrelationId(), future);
+        messageProducer.send(topic, message);
+        return future;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public CompletableFuture<List<InfoData>> sendAndGetChunked(String topic, Message message) {
+        CompletableFuture<List<InfoData>> future = new CompletableFuture<>();
+        pendingChunkedRequests.put(message.getCorrelationId(), future);
+        messageProducer.send(topic, message);
+        return future;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public void send(String topic, Message message) {
+        messageProducer.send(topic, message);
+    }
+
+    /**
+     * Processes messages that come back to NB topics, usually messages come as a response to some request.
+     * If this response is for pended request then such request will be marked as completed.
+     * Messages might be chunked and not chunked. If chunked we need to wait until we receive the last one
+     * and only then collect all responses and complete the request.
+     * @param message received message.
+     */
+    void onResponse(Message message) {
+        if (!isValid(message)) {
+            logger.warn("Received invalid message: {}", message);
+            return;
+        }
+
+        if (message instanceof ErrorMessage) {
+            ErrorMessage error = (ErrorMessage) message;
+            logger.error("Response message is error: {}", error);
+            throw new MessageException(error);
+        }
+
+        if (message instanceof ChunkedInfoMessage) {
+            processChunkedMessage((ChunkedInfoMessage) message);
+        } else if (pendingRequests.containsKey(message.getCorrelationId())) {
+            InfoMessage infoMessage = (InfoMessage) message;
+            CompletableFuture<InfoData> future = pendingRequests.remove(message.getCorrelationId());
+            future.complete(infoMessage.getData());
+        }
+    }
+
+    /**
+     * Performs searching and collecting all chunked messages into one chain if possible.
+     */
+    private synchronized void processChunkedMessage(ChunkedInfoMessage received) {
+        List<ChunkedInfoMessage> chain;
+        if (messagesChains.containsKey(received.getCorrelationId())) {
```

Comment:
We can put empty list here in sendAndGetChunked method, and remove this check.

PS we don't need to call .containsKey, just
```
chain = messagesChains.remove(key);
if (chain != null) {
    chain.add(received);
    if (received.getNextCorrelationId() != null) {
        messagesChains.put();
    } else {
        handleChainComplete();
    }
}